app/js/app/directives/graph_sketcher/GraphSketcher.js

```diff
@@ -0,0 +1,2586 @@
+"use strict";
+define(function(require) {
+    return ["$timeout", "$rootScope", "api", function($timeout, $rootScope, api) {
+        // we require instances of bezier, func and sampler to enable access to external methods.
+        var b = require('lib/graph_sketcher/bezier.js');
+        var f = require('lib/graph_sketcher/func.js');
+        var s = require('lib/graph_sketcher/sampler.js');
+        // var MySketch = require("inequality").MySketch;
+
+        var instanceCounter = 0;
+        return {
+            // scope: true,
+            // transclude: true,
+            restrict: "A",
+            templateUrl: "/partials/graph_sketcher/graph_sketcher.html",
+
+            link: function(scope, element, attrs) {
+
+                element.on("touchstart touchmove", "canvas", function(e) {
+                    e.preventDefault();
+                });
+
+                scope.title = "Sketcher";
+                scope.canvasOffset = {};
+                scope.draggingNewSymbol = false;
+                scope.equationEditorElement = element;
+
+                var colorSelect = element.find(".color-select")[0];
+                var encodeData;
+                var decodeData;
+                var reDraw;
+
+
+                scope.submit = function() {
+                    $("#graphModal").foundation("reveal", "close");
+                };
+
+                scope.logOnClose = function(event) {
+                    // This ought to catch people who navigate away without closing the editor!
+                    if (scope.log != null) {
+                        scope.log.actions.push({
+                            event: "NAVIGATE_AWAY",
+                            timestamp: Date.now()
+                        });
+                        api.logger.log(scope.log);
```

Comment:
Is this safe?

If you didn't intend to form a log event in the way `EQN_EDITOR_LOG` works, this could end badly!

Reply:
scope.log is defined at 2526 before scope.logOnClose is registered. However we only add to the initial log on "close" for now.